pkg/csi/service/driver.go

```diff
@@ -134,6 +141,61 @@ func (driver *vsphereCSIDriver) BeforeServe(ctx context.Context) error {
 			log.Errorf("failed to read config. Error: %+v", err)
 			return err
 		}
+
+		CSINamespace := os.Getenv(csitypes.EnvVarNamespace)
+		if CSINamespace == "" {
+			CSINamespace = cnsconfig.DefaultCSINamespace
+		}
+
+		if commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.CSIInternalGeneratedClusterID) &&
+			clusterFlavor == cnstypes.CnsClusterFlavorVanilla && cfg.Global.ClusterID == "" {
+			// In case of vanilla k8s deployments, if cluster ID is not provided in the
+			// vSphere config secret, then generate an unique cluster ID internally.
+			clusterID := getNewUUID()
+
+			// Create the immutable ConfigMap to store cluster ID, so that it will be
+			// persisted in etcd and it can't be updated by any user.
+			configMapData := make(map[string]string)
+			configMapData["clusterID"] = clusterID
+
+			cmData, err := commonco.ContainerOrchestratorUtility.CreateConfigMap(ctx,
+				cnsconfig.ClusterIDConfigMapName, CSINamespace,
+				configMapData, true)
+			if err != nil {
+				log.Errorf("Failed to create the immutable ConfigMap, Err: %v", err)
+				return err
+			}
+
+			if cmData != nil {
+				// If ConfigMap for cluster ID already exists, then instead of
+				// using newly generated clusterID value, we will use the
+				// clusterID value stored in the existing immutable ConfigMap.
+				clusterID = cmData["clusterID"]
+				log.Infof("clusterID is not provided in vSphere config secret, "+
+					"using the clusterID %s from existing ConfigMap", clusterID)
+			} else {
+				log.Infof("clusterID is not provided in vSphere config secret, "+
+					"generated a new clusterID %s", clusterID)
+			}
+
+			cnsconfig.GeneratedClusterID = clusterID
+			cfg.Global.ClusterID = clusterID
+		} else if clusterFlavor == cnstypes.CnsClusterFlavorVanilla && cfg.Global.ClusterID != "" {
+			// If cluster ID is provided by user in vSphere config secret and immutable
```

Comment:
This situation can only be possible if customer did not give clusterID before and later on updated the vSphere config secret right? If that's the case and you allow customer to delete the configmap and use the vSphere config provided clusterID, won't it affect the volume metadata which was already stored with the previous ID?

Reply:
This situation can only be possible if customer did not give clusterID before and later on updated the vSphere config secret right?
--> This is one case and another case could be when customer downgrades to older version by providing cluster ID in config secret and doesn't delete immutable ConfigMap. But, we are planning to document downgrade steps as mentioned in the functional spec.
If we want to rely on documentation for downgrade and care about upgrade scenario, should we log to delete cluster  ID from config secret?

Reply:
Yes, this situation will occur if customer did not give clusterID before and later on updated the vSphere config secret. Updated comment/log message to ask user to delete cluster ID from the vSphere config secret. And we will use cluster ID stored in immutable CM once this is fixed.